Lecture04

Vitald75 opened the pull request:
Lecture04. 
Add DisplayName, Order, Tag, ParametrizedTest notations, TestSuite, 
Add soft and hard asserts.

Add new test classes.

Vitald75 committed:
Add methods (create locators and flow): testAddRecruitmentCandidate,
testAddJobTitle

VitaliDauhashei committed:
Minor changes

Vitald75 committed:
Add some locators to the testAddRecruitmentCandidate,

Vitald75 committed:
Refactoring. Create BaseTest Class,LoginPage Class

VitaliDauhashei committed:
Minor changes

Vitald75 committed:
Refactoring. Add LogoutTest class, add asserts into LoginPageTest and LogoutTest classes, add @DisplayName,@Tag,

Vitald75 committed:
Order tests, parametrized tests, DisplayName.
Use Tags, testSuite.
Add AddRecruitmentCandidateTest class,
Add AddDeleteJobTitlesTest class,
Add soft and asserts.

DzianisShapel commented on src/test/java/eu/senla/LoginPageTest.java:
```diff
@@ -0,0 +1,70 @@
+  @Test
+  @Order(0)
+  @DisplayName("Проверка успешного логина")
+  @Tag("smoke")
```
ты будешь пытаться запускать по тэгам или по сьютам?

Vitald75 replied:
На данный момент при текущей организации кейсов более выгодным выглядит запуск по Тэгам, (для чего надо убрать будет загасить созданные suite, либо внутри файла сьюта оставить включенным один класс, не содержащий тэгов, тогда не будет дублированного прогона кейсов.

пока созданный сьют не удаляю, для возможности дальнейших экспериментов

DzianisShapel commented on src/test/java/eu/senla/BaseTest.java:
```diff
@@ -0,0 +1,42 @@
+    driver.manage().deleteAllCookies();
+    driver.get("https://opensource-demo.orangehrmlive.com/web/index.php/auth/login");
+    final int waitingTime = 6;
+    Wait<WebDriver> wait = new WebDriverWait(driver, Duration.ofSeconds(waitingTime));
```
у тебя есть базовый класс в котором переменная на уровне класса protected static WebDriver driver; 
а инициализируется она в методе 
`@BeforeAll
  static void init() {
    driver = new ChromeDriver();
  }`
по такой же аналогии ты можешь объявить переменную Wait<WebDriver> wait на уровне класса 
а иницилиазировать также в 
@BeforeAll
  static void init() {
    driver = new ChromeDriver();
wait = new WebDriverWait(driver, Duration.ofSeconds(waitingTime)); 
  } 
  и тогда все саб-классы (тесты) унаследуют этот экземпляра wait 
  и не надо будет в тесте каждый раз создавать

Vitald75 replied:
Собирался так и сделать, откладывал на потом, так и отложилось...
Вынес в базовый класс

DzianisShapel commented on src/test/java/eu/senla/LoginPageTest.java:
```diff
@@ -0,0 +1,70 @@
+    wait.until(
+        d -> driver.findElement(By.xpath("//p[text()='Invalid credentials']")).isDisplayed());
+
+    assertTrue(
```
это какие ассершны хард или софт?

Vitald75 replied:
это хард , 
софт применен в классе LogoutTest :
assertAll ( ()-> assertTrue(),
                .....
                ()-> assertEquals(),
              )

DzianisShapel commented on src/test/java/eu/senla/BaseTest.java:
```diff
@@ -0,0 +1,42 @@
+public class BaseTest {
+  protected static WebDriver driver;
+
+  @BeforeAll
```
Start each test from a clean, known state. Ideally, spin up a new virtual machine for each test. If spinning up a new virtual machine is not practical, at least start a new WebDriver for each test. Most browser drivers like GeckoDriver and ChromeDriver will start with a clean known state with a new user profile, by default. 
https://www.selenium.dev/documentation/test_practices/encouraged/fresh_browser_per_test/ 
поэтому я бы сделал BeforEach

Vitald75 replied:
Done

DzianisShapel commented on src/test/java/eu/senla/BaseTest.java:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  public final void successfulLogin() {
+    driver.manage().deleteAllCookies();
```
и есть бифор ич тогда вот это наверно не надо

Vitald75 replied:
Действительно, использование @BeforeEach решило проблему сохраненной сессии(кредов), 
перевел на BeforeEach, убрал удаление Cookies.

DzianisShapel commented on src/test/java/eu/senla/AppTest.java:
```diff
@@ -1,71 +1,56 @@
-  Wait<WebDriver> wait = new WebDriverWait(driver, Duration.ofSeconds(2));
-  wait.until(d -> driver.findElement(By.name("username"))
-          .isDisplayed());
+public class AppTest extends BaseTest {
```
переименуй его во что-нибудь вменяемое

Vitald75 replied:
Переименовал.

DzianisShapel commented on src/test/java/eu/senla/AppTest.java:
```diff
@@ -1,71 +1,56 @@
 
     // open PIM page
     driver.findElement(By.cssSelector("a[href$='viewPimModule']")).click();
     driver.manage().window().maximize();
```
лучше вот эту штуку вынеси с базовый класс и в одном месте будешь растягивать экран для всех тестов

DzianisShapel commented on src/test/java/eu/senla/AppTest.java:
```diff
@@ -1,71 +1,56 @@
 
     // fill form and click Save button
     driver.findElement(By.name("firstName")).sendKeys("Alexandr");
     driver.findElement(By.name("middleName")).sendKeys("Sergeevich");
```
на данном этапе сойдет, но это очень хардкодно ))) 
после следующей лекции переделаем

DzianisShapel commented on src/test/java/eu/senla/AppTest.java:
```diff
@@ -1,71 +1,56 @@
 
-    wait.until(d -> driver.findElement(By.xpath("//h6[text()='Alexandr PushyKin']")))
-            .isDisplayed();
+    wait.until(d -> driver.findElement(By.xpath("//h6[text()='Alexandr PushyKin']"))).isDisplayed();
```
на данном этапе пойдет 
но в целом такие локаторы не катят 
на лекции объясню почему

DzianisShapel commented on config/checkstyle-suppressions.xml:
```diff
@@ -6,8 +6,13 @@
+    <suppress checks="JavadocVariable" files=""/>
     <suppress checks="FinalParameters" files=""/>
     <suppress checks="HideUtilityClassConstructor" files=""/>
+    <suppress checks="AvoidStarImport" files=""/>
```
а вот это правило классное кстати 
не допускает множественный импорт  
но исключил, так исключил 
в след потоке студентов - будет обязательным  ))))

Vitald75 replied:
есть еще одно правило, которое я пока исключил - про ограничение количества параметров в методе не более 7, 
я для заполнения формы передаю 8, можно конечно некоторые необязательные поля захардкодить,
// вопрос для обсуждения  

а про * - посмотрю, что можно сделать)

DzianisShapel commented on pom.xml:
```diff
@@ -13,10 +13,13 @@
-    <selenium.version>4.31.0</selenium.version>
+    <selenium.version>4.32.0</selenium.version>
+    <common.lang.version>2.6</common.lang.version>
+    <groups>smoke,extended</groups>
```
вот это наверно лишнее 
ты в плагине сделал вот так 
```
        <configuration>
          <groups>${groups}</groups>
        </configuration>  
```
 если убрать 22 строку, то в плагине эта переменка красная 
 но ничего страшного 
 если запускать 
 mvn clean test -Dgroups=smoke 
 то все будет работать 
 мавен сам из терминальной команды подставит это значение 
 так получается более гибко

DzianisShapel replied:
это все делается на перспективу, чтоб связать все с CI 
в будущем будет Jenkins джоба параметризованная, у которой будет выпадашка со значениями всех тэгов 
при запуске джобы ты выберешь конкретный тэг
дженкис подставит и сформирует команду 
`mvn clean test -Dgroups=smoke `
и запустит

Vitald75 replied:
То, что красная переменка в плагине - это факт, 
поэтому и добавлял groups, 
убрал  -  действительно работает и без явного указания значения переменной

VitaliDauhashei committed:
Minor changes after review by Dzianis.
Rename class AppTest, use BeforeEach instead of BeforeAll, etc

Vitald75 committed:
Order tests, parametrized tests, DisplayName.
Use Tags, testSuite.
Add AddRecruitmentCandidateTest class,
Add AddDeleteJobTitlesTest class,
Add soft and asserts.
Resolve .* problem in imports.
Modify LoginPageTest, add case with empty username and password

Vitald75 committed:
Order tests, parametrized tests, DisplayName.
Use Tags, testSuite.
Add AddRecruitmentCandidateTest class,
Add AddDeleteJobTitlesTest class,
Add soft and asserts.
Resolve .* problem in imports.
Modify LoginPageTest, add case with empty username and password
